# Prueba Técnica DT DataScience

Comenzamos el proyecto importando las librerías necesarias y echando un primer vistazo a los datos disponibles.

In [36]:
import pandas as pd
import numpy as np


In [3]:
data_2002 = pd.read_csv('Data/Censo_2002.csv', sep=',')
data_2018 = pd.read_csv('Data/Censo_2018.csv', sep=',')

In [4]:
data_2002.head()

COD_DANE  Total_Viviendas  Casas  Casas_indigenas  \
0  7300110000000000590101               37     23              NaN   
1  7300110000000000320703               30     16              NaN   
2  7300110000000000330935              156     35              NaN   
3  7300110000000000330221               15     14              NaN   
4  7300110000000000330103               36     34              NaN   

   Apartamento  Tipo_Cuarto  Electrica_SI  Electrica_NO  GasNatural_SI  \
0           14            0            34             0             30   
1           14            0            30             0             28   
2          117            3           146             1             89   
3            1            0            14             1              7   
4            2            0            36             0             32   

   GasNatur_NO  ...  Total_Hogares  Total_Personas  HOMBRE  MUJER  Preescolar  \
0            4  ...             35             146      59     87           2   
1            2  ...             34             129      51     78           3   
2           58  ...            149             511     243    268          17   
3            8  ...             15              48      27     21           0   
4            4  ...             36             115      52     63           2   

   Primaria  Secundaria  Especializacion  Ninguna_Educacion  No_responde  
0        44          21                1                 15           63  
1        22          29                0                  5           70  
2        97          88                4                 30          275  
3         6           9                1                  0           32  
4        16          13                2                  2           80  

[5 rows x 24 columns]

In [5]:
data_2018.head()

COD_DANE_A  VERSION         AREA  Casa_indigena  \
0  7300110000000000000000     2018  5379.862696              0   
1  7300110000000000130204     2018  3244.940412              0   
2  7300110000000000130206     2018  2081.423166              0   
3  7300110000000000130207     2018  5619.115851              0   
4  7300110000000000130208     2018  4902.728507              0   

   Total_Viviendas  Casa  Apartamento  Tipo_Cuarto  Electrica_SI  \
0               30    12           18            0            46   
1               34    32            2            0            56   
2               24    15            9            0            30   
3               65    59            4            2           112   
4               56    44           11            1            94   

   Electrica_NO  ...  Total_Personas  HOMBRE  MUJER  Preescolar  Primaria  \
0             0  ...              87      45     42          18         0   
1             0  ...              84      40     44           0         0   
2             0  ...              38      18     20           0         0   
3             0  ...             177      72    105          14         0   
4             0  ...             153      70     83          30         0   

   Secundaria  Especializacion  Ninguno  Ninguno.1  No_responde  
0          24                0        1          2           60  
1          40                0        1          0           43  
2          18                0        1          0           19  
3          84                0        1          3           89  
4          44                0        1          6          102  

[5 rows x 26 columns]

In [6]:
data_2002.describe()

Total_Viviendas       Casas  Casas_indigenas  Apartamento  Tipo_Cuarto  \
count       574.000000  574.000000              0.0   574.000000   574.000000   
mean         31.297909   21.794425              NaN     8.290941     1.013937   
std          36.292444   24.748429              NaN    17.245604     2.531439   
min           0.000000    0.000000              NaN     0.000000     0.000000   
25%          16.000000    8.000000              NaN     1.000000     0.000000   
50%          26.000000   19.000000              NaN     4.000000     0.000000   
75%          38.000000   28.000000              NaN     9.000000     1.000000   
max         484.000000  346.000000              NaN   275.000000    40.000000   

       Electrica_SI  Electrica_NO  GasNatural_SI  GasNatur_NO  \
count    574.000000    574.000000     574.000000   574.000000   
mean      30.167247      0.304878      23.388502     6.693380   
std       34.273715      2.635390      26.716283    11.329896   
min        0.000000      0.000000       0.000000     0.000000   
25%       15.000000      0.000000      11.000000     2.000000   
50%       25.000000      0.000000      19.000000     4.000000   
75%       37.000000      0.000000      30.000000     8.000000   
max      453.000000     54.000000     323.000000   154.000000   

       Alcantarillado_SI  ...  Total_Hogares  Total_Personas      HOMBRE  \
count         574.000000  ...     574.000000      574.000000  574.000000   
mean           29.895470  ...      32.639373      116.825784   54.555749   
std            33.886829  ...      38.517246      131.479038   61.771997   
min             0.000000  ...       0.000000        0.000000    0.000000   
25%            15.000000  ...      16.000000       56.000000   26.000000   
50%            25.000000  ...      26.000000      100.000000   46.000000   
75%            37.000000  ...      39.000000      146.000000   68.000000   
max           450.000000  ...     538.000000     1850.000000  861.000000   

            MUJER  Preescolar    Primaria  Secundaria  Especializacion  \
count  574.000000  574.000000  574.000000  574.000000       574.000000   
mean    62.270035    3.405923   32.839721   23.289199         1.421603   
std     69.997745    5.040918   41.417705   26.763002         3.440842   
min      0.000000    0.000000    0.000000    0.000000         0.000000   
25%     30.000000    1.000000   13.000000   10.000000         0.000000   
50%     52.500000    2.000000   27.000000   19.000000         0.000000   
75%     78.000000    4.000000   42.000000   30.000000         2.000000   
max    989.000000   60.000000  685.000000  400.000000        50.000000   

       Ninguna_Educacion  No_responde  
count         574.000000   574.000000  
mean            7.095819    48.773519  
std            11.231479    58.690250  
min             0.000000     0.000000  
25%             1.000000    18.000000  
50%             5.000000    38.000000  
75%             9.000000    62.750000  
max           189.000000   727.000000  

[8 rows x 23 columns]

In [7]:
data_2018.describe()

VERSION           AREA  Casa_indigena  Total_Viviendas        Casa  \
count    618.0     618.000000          618.0       618.000000  618.000000   
mean    2018.0    6241.777085            0.0        36.556634   22.762136   
std        0.0   31672.296709            0.0        64.740331   38.402796   
min     2018.0      58.290387            0.0         0.000000    0.000000   
25%     2018.0    1153.578599            0.0        15.000000    8.000000   
50%     2018.0    1902.007629            0.0        27.000000   18.000000   
75%     2018.0    3327.372506            0.0        39.750000   28.750000   
max     2018.0  525251.809958            0.0      1075.000000  809.000000   

       Apartamento  Tipo_Cuarto  Electrica_SI  Electrica_NO  Acueducto_SI  \
count   618.000000   618.000000    618.000000    618.000000    618.000000   
mean     12.953074     0.786408     63.420712      0.082524     31.537217   
std      44.955906     2.332147    112.575612      1.033312     56.039082   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.000000     24.000000      0.000000     12.000000   
50%       3.000000     0.000000     48.000000      0.000000     24.000000   
75%       9.000000     1.000000     70.000000      0.000000     35.000000   
max     701.000000    33.000000   2120.000000     25.000000   1050.000000   

       ...  Total_Personas       HOMBRE        MUJER  Preescolar    Primaria  \
count  ...      618.000000   618.000000   618.000000  618.000000  618.000000   
mean   ...       94.773463    44.802589    49.970874   17.135922   12.161812   
std    ...      142.257524    72.728775    70.285104   27.561118   49.838374   
min    ...        0.000000     0.000000     0.000000    0.000000    0.000000   
25%    ...       38.000000    17.250000    20.000000    0.000000    0.000000   
50%    ...       73.000000    34.000000    39.000000   10.000000    0.000000   
75%    ...      109.750000    53.000000    58.750000   26.000000    0.000000   
max    ...     2379.000000  1324.000000  1055.000000  384.000000  922.000000   

       Secundaria  Especializacion     Ninguno   Ninguno.1  No_responde  
count  618.000000       618.000000  618.000000  618.000000   618.000000  
mean    22.610032         1.744337    1.786408    1.472492    54.998382  
std     46.873377        11.534432    2.882858    5.691706    53.305593  
min      0.000000         0.000000    0.000000    0.000000     0.000000  
25%      0.000000         0.000000    0.000000    0.000000    25.000000  
50%     13.000000         0.000000    1.000000    0.000000    47.000000  
75%     28.000000         0.000000    3.000000    1.000000    73.000000  
max    661.000000       166.000000   43.000000  121.000000   625.000000  

[8 rows x 25 columns]

## Pregunta 1

### Cuál es el porcentaje (rata) de aparición de nuevas manzanas cada 3 años

Para obtener el resultado de la pregunta 1 empezaremos por contar la cantidad de manzanas que hay cada año.

In [8]:
data_2002.COD_DANE.count()

574

In [9]:
data_2018.COD_DANE_A.count()

618

Calculando la diferencia entre la cantidad de manzanas que han aparecido entre los años 2002 y 2018 vemos que hay 44 manzanas nuevas. Consecuentemente podemos calcular la cantidad de manzanas que han aparecido al año:

$$\frac{44\> \mathrm{manzanas}}{16\> \mathrm{años}}= 2,75 \>\mathrm{manzanas\>al\>año}$$

De esta manera podemos calcular que en tres años hay un incremento de 8,25 manzanas. Para calcular el porcentaje de crecimiento utilizaremos una regla de tres: $$\frac{8,25}{574\> \mathrm{manzanas}}\times 100 = 1,44\% $$

## Pregunta 2

### Cuál es la proyección de hombres por hogar al año 2025

Para comenzar sumaremos la cantidad de hombres que había en cada año utilizando el comando <code>.sum()</code>

In [10]:
data_2002.HOMBRE.sum()

31315

In [11]:
data_2018.HOMBRE.sum()

27688

Despueés procederemos a calcular la tasa de crecimiento de los hombres entre estos años: $$\frac{27688-31315}{2018-2002}=-226,68\> \mathrm{hombres\> por \> año}$$ De esa manera podemos obtener una ecuación lineal que nos permitirá saber la población de hombres que habrá para el año 2025. $$y=27688-226,68\times (2025-2018) \approx 26101.$$ Para el año 2025 esperamos tener una reducción de hombres hasta una población de 26101.

## Pregunta 3

### Cuál es el porcentaje de aumento de escolaridad por unidad de área para las categorías de preescolar y especialización

Comenzaremos por crear dos listas distintas. En una de ellas guardaremos el aumento de escolaridad por unidad de área para aquellas manzanas que ya existían en 2002 y en otra aquellas que han sido contruidas durante ese periodo de tiempo. Después juntaremos las dos listas en una última lista que tendrá el porcentaje de aumento de cada manzana.

In [29]:
aumento_lista_existente_preescolar = []
aumento_lista_existente_preescolar = [(data_2018.loc[data_2018.COD_DANE_A == manzana, 'Preescolar'].iloc[0]-data_2002.loc[data_2002.COD_DANE == manzana, 'Preescolar'].iloc[0])*100/data_2018.loc[data_2018.COD_DANE_A == manzana, 'AREA'].iloc[0] for manzana in data_2018.COD_DANE_A if manzana not in aumento_lista_existente_preescolar and manzana in data_2002.COD_DANE]
aumento_lista_nueva_preescolar = [(data_2018.loc[data_2018.COD_DANE_A == manzana, 'Preescolar'].iloc[0])*100/data_2018.loc[data_2018.COD_DANE_A == manzana, 'AREA'].iloc[0] for manzana in data_2018.COD_DANE_A if manzana not in aumento_lista_existente_preescolar and manzana not in data_2002.COD_DANE]
aumento_lista_preescolar = [*aumento_lista_existente_preescolar, *aumento_lista_nueva_preescolar]

Guardaremos estas listas en DataFrames ya que nos permiten tener una visualización más clara de la información y nos conceden la oportunidad de utilizar métodos que nos serán necesarios más adelante.

In [30]:
aumento_preescolar = pd.DataFrame(aumento_lista_preescolar, columns=['Porcentaje de Aumento por Area (Preescolar)'])
aumento_preescolar.head()

Porcentaje de Aumento por Area (Preescolar)
0                                     0.334581
1                                     0.000000
2                                     0.000000
3                                     0.249150
4                                     0.611904

Repetimos el proceso por cada categoría.

In [31]:
aumento_lista_existente_especializacion = []
aumento_lista_existente_especializacion = [(data_2018.loc[data_2018.COD_DANE_A == manzana, 'Especializacion'].iloc[0]-data_2002.loc[data_2002.COD_DANE == manzana, 'Especializacion'].iloc[0])*100/data_2018.loc[data_2018.COD_DANE_A == manzana, 'AREA'].iloc[0] for manzana in data_2018.COD_DANE_A if manzana not in aumento_lista_existente_especializacion and manzana in data_2002.COD_DANE]
aumento_lista_nueva_especializacion = [(data_2018.loc[data_2018.COD_DANE_A == manzana, 'Especializacion'].iloc[0])*100/data_2018.loc[data_2018.COD_DANE_A == manzana, 'AREA'].iloc[0] for manzana in data_2018.COD_DANE_A if manzana not in aumento_lista_existente_especializacion and manzana not in data_2002.COD_DANE]
aumento_lista_especializacion = [*aumento_lista_existente_especializacion, *aumento_lista_nueva_especializacion]

In [34]:
aumento_especializacion = pd.DataFrame(aumento_lista_especializacion, columns=['Porcentaje de Aumento por Area (Especializacion)'])
aumento_especializacion

Porcentaje de Aumento por Area (Especializacion)
0                                                 0.0
1                                                 0.0
2                                                 0.0
3                                                 0.0
4                                                 0.0
..                                                ...
613                                               0.0
614                                               0.0
615                                               0.0
616                                               0.0
617                                               0.0

[618 rows x 1 columns]

Checkeando la cantidad de valores diferentes a 0 en el DataFrame.

In [37]:
print(np.count_nonzero(aumento_especializacion))

40


Finalmente para obtener el porcentaje de aumento de escolaridad por unidad de área para las categorías de preescolar y especialización calcularemos la media de todos los valores. Para ello podemos utilizar el metodo <code> .describe() </code> ya que nos ofrece este dato directamente.

In [38]:
aumento_preescolar.describe()

Porcentaje de Aumento por Area (Preescolar)
count                                   618.000000
mean                                      0.702652
std                                       0.876323
min                                       0.000000
25%                                       0.000000
50%                                       0.362699
75%                                       1.284107
max                                       6.112752

In [39]:
aumento_especializacion.describe()

Porcentaje de Aumento por Area (Especializacion)
count                                        618.000000
mean                                           0.024358
std                                            0.125948
min                                            0.000000
25%                                            0.000000
50%                                            0.000000
75%                                            0.000000
max                                            1.830642

Analizando los datos vemos que el aumento de escolarización por unidad de área para la categoría de Preescolar es del 70,2%. Repitiendo el mismo porceso para el caso de Especialización tenemos que el aumento es del 2,43%.

Por otro lado, debajo se deja una función general que nos permite calcular todo lo anterior sin necesidad de escribir todo el código de nuevo:

In [44]:
def aumento_escolaridad(data_vieja, data_nueva, categoria):
    """Función de cálculo para el aumento_escolaridad por categorías.

    Args:
        data_vieja (DataFrame): DataFrame conteniendo datos del año más viejo
        data_nueva (DataFrame): DataFrame conteniendo datos del año más nuevo
        categoria (str): Categoría que se quiere analizar

    Returns:
        aumento [DataFrame]: Devuelve un DataFrame con el aumento de escolarización por unidad de área para la categoría seleccionada.
    """

    aumento_lista_existente = []
    aumento_lista_existente = [(data_nueva.loc[data_nueva.COD_DANE_A == manzana, categoria].iloc[0]-data_vieja.loc[data_vieja.COD_DANE == manzana, categoria].iloc[0])*100/data_nueva.loc[data_nueva.COD_DANE_A == manzana, 'AREA'].iloc[0] for manzana in data_nueva.COD_DANE_A if manzana not in aumento_lista_existente and manzana in data_vieja.COD_DANE]
    aumento_lista_nueva = [(data_nueva.loc[data_nueva.COD_DANE_A == manzana, categoria].iloc[0])*100/data_nueva.loc[data_nueva.COD_DANE_A == manzana, 'AREA'].iloc[0] for manzana in data_nueva.COD_DANE_A if manzana not in aumento_lista_existente and manzana not in data_vieja.COD_DANE]
    aumento_lista = [*aumento_lista_existente, *aumento_lista_nueva]
    aumento = pd.DataFrame(aumento_lista, columns=['Porcentaje de Aumento por Area ('+ categoria +')'])
    return aumento

Comprobamos el correcto funcionamiento con la categoría de Secundaria:

In [45]:
aumento_secundaria = aumento_escolaridad(data_2002, data_2018, 'Secundaria')

In [46]:
aumento_secundaria.describe()

Porcentaje de Aumento por Area (Secundaria)
count                                   618.000000
mean                                      0.761449
std                                       0.829247
min                                       0.000000
25%                                       0.000000
50%                                       0.711170
75%                                       1.208671
max                                      10.522211

## Pregunta 4